views/clickhouse.go

```diff
@@ -0,0 +1,141 @@
+package views
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"github.com/ClickHouse/clickhouse-go/v2"
+	"github.com/ClickHouse/clickhouse-go/v2/lib/driver"
+	"github.com/Masterminds/squirrel"
+	"strings"
+	"time"
+)
+
+const maxClickhouseResultRows = 1000
+
+type RealtimeViewershipRow struct {
+	Timestamp   time.Time `ch:"timestamp_ts"`
+	ViewCount   uint64    `ch:"view_count"`
+	BufferRatio float64   `ch:"buffer_ratio"`
+	ErrorRate   float64   `ch:"error_rate"`
+
+	PlaybackID  string `ch:"playback_id"`
+	DeviceType  string `ch:"device_type"`
+	Browser     string `ch:"browser"`
+	CountryName string `ch:"playback_country_name"`
+}
+
+type Clickhouse interface {
+	QueryRealtimeViewsEvents(ctx context.Context, spec QuerySpec) ([]RealtimeViewershipRow, error)
+}
+
+type ClickhouseOptions struct {
+	Addr     string
+	User     string
+	Password string
+	Database string
+}
+
+type ClickhouseClient struct {
+	conn driver.Conn
+}
+
+func NewClickhouseConn(opts ClickhouseOptions) (*ClickhouseClient, error) {
+	conn, err := clickhouse.Open(&clickhouse.Options{
+		Addr: strings.Split(opts.Addr, ","),
+		Auth: clickhouse.Auth{
+			Database: opts.Database,
+			Username: opts.User,
+			Password: opts.Password,
+		},
+		TLS: &tls.Config{},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &ClickhouseClient{conn: conn}, nil
+}
+
+func (c *ClickhouseClient) QueryRealtimeViewsEvents(ctx context.Context, spec QuerySpec) ([]RealtimeViewershipRow, error) {
+	sql, args, err := buildRealtimeViewsEventsQuery(spec)
+	if err != nil {
+		return nil, fmt.Errorf("error building viewership events query: %w", err)
+	}
+	var res []RealtimeViewershipRow
+	err = c.conn.Select(ctx, &res, sql, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	if spec.From == nil && spec.To == nil {
+		return filterMostRecent(res), nil
```

Comment:
In https://www.notion.so/livepeer/Viewership-Queries-e27bdb9df64240cea371ca4e0f11c0f6?pvs=4, 

I designed a separate query for the "Current Totals" query. I'll need to evaluate how this can differ. 

I'm most concerned about the first 5 to 10 seconds of the minute. `timestamp_ts` is the minute rounded down so, for those first few seconds the data may be incomplete as heartbeats come in and we factor in latency.

Reply:
Ahh, I didn't notice you created a separate query for the "currently active'. Sorry for that.

But I wonder, in general, why would we have a different query for the "current view count" and the "last 30 current view counts"? It sounds counterintuitive to query this differently.

I think you may be right that there may be a problem with the first few seconds of the minute, but maybe then we should not use the last minute at all? Just always take the data from the minute before? I think we should not sacrifice the data correctness for the latency. No matter if we query "last 30 min" or the current view count. Wdyt?

Reply:
>  It sounds counterintuitive to query this differently.
They are different questions. The by-minute query is grouped by minute whereas the "current totals" ignores the clock minute and is only interested in the last 30 seconds. 

Also, the `viewership_sessions_by_minute` is a rollup and optimization on top of the `viewership_events` table. its optimized specifically for the "by minute" question. Its not meant to answer the last 30 seconds question.

> but maybe then we should not use the last minute at all? 
I was actually already thinking about this too. Lets do it but only for the by-minute. But we should also bump up the "current totals" to be the last 60 seconds. Two benefits, one: we get a little more flexibility with latency and two: this 60 seconds will fill in the gap we are "dropping" from the by-minute query. 

wdyt?

Reply:
ok, it makes sense.

For "not querying the last 1 minute", here's the change: bdbe7893fe49f05789d053f566163de6c82bb8c7 . PTAL